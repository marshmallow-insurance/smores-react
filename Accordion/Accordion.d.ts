import { FC, ReactNode } from 'react';
import { MarginProps } from '../utils/space';
export type AccordionProps = {
    title: string;
    subTitle?: string;
    filledBackground?: boolean;
    borderTop?: boolean;
    borderColor?: 'oatmeal' | 'custard' | 'cream' | 'coconut';
    fullBorder?: boolean;
    backgroundColor?: 'oatmeal' | 'custard' | 'cream' | 'coconut';
    onToggle?: (isOpen: boolean) => void;
    children: ReactNode;
    defaultIsOpen?: boolean;
} & MarginProps;
export declare const Accordion: FC<AccordionProps>;
