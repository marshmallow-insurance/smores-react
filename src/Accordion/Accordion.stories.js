import React from 'react'
import { Accordion } from './Accordion'

export default {
  title: 'Accordion',
  component: Accordion,
}

const Template = (args) => (
  <Accordion {...args} title="How it works">
    Lots of brilliant information about this beautiful component
  </Accordion>
)

export const Default = Template.bind({})
