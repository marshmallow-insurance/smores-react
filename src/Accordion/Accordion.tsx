import React, { FC, useState } from 'react'
import styled, { css } from 'styled-components'

import { Box } from '../Box'
import { Icon } from '../Icon'
import { theme } from '../theme'
import { Text } from '../Text'
import { MarginProps } from '../utils/space'

export type AccordionProps = {
  title: string
  subTitle?: string
  borderTop?: boolean
  fullBorder?: boolean
  onToggle?: VoidFunction
} & MarginProps

export const Accordion: FC<AccordionProps> = ({
  title,
  children,
  onToggle,
  borderTop = false,
  subTitle,
  fullBorder = false,
  ...marginProps
}) => {
  const [isOpen, setIsOpen] = useState(false)
  const px = fullBorder ? '16px' : '0'

  const handleToggle = () => {
    onToggle?.()
    setIsOpen(!isOpen)
  }

  return (
    <Wrapper borderTop={borderTop} fullBorder={fullBorder} {...marginProps}>
      <TopContainer
        flex
        alignItems="center"
        justifyContent="space-between"
        onClick={handleToggle}
        py={fullBorder ? { custom: 20 } : '16px'}
        px={px}
      >
        <TitleContainer>
          <Text tag="h2" typo="headline-regular">
            {title}
          </Text>
          {subTitle && (
            <Text tag="label" color="subtext" typo="label">
              {subTitle}
            </Text>
          )}
        </TitleContainer>

        <CaretIcon
          render="caret"
          size={24}
          color="secondary"
          isOpen={isOpen}
          borderTop={borderTop}
        />
      </TopContainer>
      {isOpen && (
        <Box pt="12px" pb="16px" px={px}>
          {children}
        </Box>
      )}
    </Wrapper>
  )
}

interface IAccordion {
  isOpen: boolean
  borderTop?: boolean
  fullBorder?: boolean
}

const Wrapper = styled(Box)<Omit<IAccordion, 'isOpen'>>(
  ({ borderTop, fullBorder }) => css`
    border-bottom: 1px solid ${theme.colors.outline};
    ${borderTop && `border-top: 1px solid ${theme.colors.outline};`}

    ${fullBorder &&
    css`
      border: 1px solid ${theme.colors.outline};
      border-radius: 8px;
      margin-bottom: 14px;
    `}
  `,
)

const TitleContainer = styled.div`
  display: flex;
  flex-direction: column;
`
const TopContainer = styled(Box)`
  cursor: pointer;
`

const CaretIcon = styled(Icon)<IAccordion>(
  ({ isOpen }) => css`
    transform: ${isOpen ? 'rotate(180deg)' : 'rotate(0deg)'};
    transition: transform 0.6s ease;
  `,
)
