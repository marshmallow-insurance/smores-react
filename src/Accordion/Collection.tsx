import React, { FC, ReactNode } from 'react'
import styled from 'styled-components'

import { Text } from '../Text'
import { Box } from '../Box'
import { Accordion, AccordionProps } from './Accordion'

const accordionList: Array<AccordionProps> = [
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
  },
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
    subTitle: 'subtitle',
  },
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
    filledBackground: true,
  },
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
    subTitle: 'subtitle',
    filledBackground: true,
  },
]

export const CollectionPage: FC = () => {
  return (
    <Wrapper>
      <Section label="Default">
        {accordionList.map((props, index) => (
          <Accordion key={index} {...props} />
        ))}
      </Section>
      <Section label="Border Top">
        {accordionList.map((props, index) => (
          <Accordion key={index} {...props} borderTop={true} />
        ))}
      </Section>
      <Section label="Full Border">
        {accordionList.map((props, index) => (
          <Accordion key={index} {...props} fullBorder={true} />
        ))}
      </Section>
    </Wrapper>
  )
}

const Wrapper = styled(Box)`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 40px;
  width: 100%;
`

const Section: FC<{ label: string; children?: ReactNode }> = ({
  label,
  children,
}) => {
  return (
    <Box width="100%" mb="16px">
      <Box mb="16px">
        <Text tag="span" typo="header-small">
          {label}
        </Text>
      </Box>
      <Column>{children}</Column>
    </Box>
  )
}

const Column = styled(Box)`
  display: flex;
  flex-direction: column;
  gap: 16px;
`
