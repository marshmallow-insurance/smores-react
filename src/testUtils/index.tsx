import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
