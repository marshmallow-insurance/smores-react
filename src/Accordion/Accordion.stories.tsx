import React from 'react'
import { Accordion, AccordionProps } from './Accordion'
import { CollectionPage } from './Collection'

export default {
  title: 'Accordion',
  component: Accordion,
}

const Template = (props: AccordionProps) => (
  <Accordion {...props} title="How it works">
    Lots of brilliant information about this beautiful component
  </Accordion>
)

export const Default = Template.bind({})

export const Collection = CollectionPage.bind({})
