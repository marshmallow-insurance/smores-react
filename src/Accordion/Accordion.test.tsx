import { expect, it } from 'vitest'
import { render } from '../testUtils'

import { Accordion } from './Accordion'
import React from 'react'

describe('Accordion', () => {
  it('renders correctly when closed', () => {
    const { container } = render(
      <Accordion title="How it works">
        Lots of brilliant information about this beautiful component
      </Accordion>,
    )
    expect(container).toMatchSnapshot()
  })
  it('renders correctly when expanded', () => {
    const { container } = render(
      <Accordion title="How it works" defaultIsOpen>
        Lots of brilliant information about this beautiful component
      </Accordion>,
    )
    expect(container).toMatchSnapshot()
  })
  it('renders correctly for fullBorder and filledBackground variant', () => {
    const { container } = render(
      <Accordion title="How it works" fullBorder filledBackground defaultIsOpen>
        Lots of brilliant information about this beautiful component
      </Accordion>,
    )
    expect(container).toMatchSnapshot()
  })
})
