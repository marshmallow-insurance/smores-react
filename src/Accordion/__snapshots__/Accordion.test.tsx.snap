// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion > renders correctly for fullBorder and filledBackground variant 1`] = `
<div>
  <div
    class="sc-bRKDuR bHjpxP sc-jwTyAe emHXsE"
  >
    <div
      class="sc-bRKDuR gwwejK sc-jJLAfE iWiHKP"
    >
      <div
        class="sc-hjsuWn hLwqBO"
      >
        <h2
          class="sc-bRKDuR bHjpxP sc-dTvVRJ jMSukC"
          cursor="inherit"
          title=""
        >
          How it works
        </h2>
      </div>
      <span
        class="sc-bRKDuR bHjpxP sc-fhHczv cTkdi sc-hwkwBN fqsjph"
        color="marzipan"
        data-testid="caret-container"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="sc-bRKDuR enIEuK"
    >
      Lots of brilliant information about this beautiful component
    </div>
  </div>
</div>
`;

exports[`Accordion > renders correctly when closed 1`] = `
<div>
  <div
    class="sc-bRKDuR bHjpxP sc-jwTyAe dRsXQD"
  >
    <div
      class="sc-bRKDuR jGGaUV sc-jJLAfE iWiHKP"
    >
      <div
        class="sc-hjsuWn hLwqBO"
      >
        <h2
          class="sc-bRKDuR bHjpxP sc-dTvVRJ jMSukC"
          cursor="inherit"
          title=""
        >
          How it works
        </h2>
      </div>
      <span
        class="sc-bRKDuR bHjpxP sc-fhHczv cTkdi sc-hwkwBN hNMPrg"
        color="marzipan"
        data-testid="caret-container"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Accordion > renders correctly when expanded 1`] = `
<div>
  <div
    class="sc-bRKDuR bHjpxP sc-jwTyAe dRsXQD"
  >
    <div
      class="sc-bRKDuR jGGaUV sc-jJLAfE iWiHKP"
    >
      <div
        class="sc-hjsuWn hLwqBO"
      >
        <h2
          class="sc-bRKDuR bHjpxP sc-dTvVRJ jMSukC"
          cursor="inherit"
          title=""
        >
          How it works
        </h2>
      </div>
      <span
        class="sc-bRKDuR bHjpxP sc-fhHczv cTkdi sc-hwkwBN fqsjph"
        color="marzipan"
        data-testid="caret-container"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="sc-bRKDuR hANJi"
    >
      Lots of brilliant information about this beautiful component
    </div>
  </div>
</div>
`;
